### PR TITLE
vw jni 8.1.0 closes #57

### LINKED ISSUE #57: Integrate new version of VW-JNI
A new release of VW should be coming shortly.  This needs to be integrated into Aloha.

### PR DESCRIPTION
**Pull request by deaktator**
This PR uses the changes from #54 but deletes the `aloha-h2o` submodule and incorporates changes from @jmorra's 57-new-vw-jni branch.  This PR:
1. Allows `vw.learner.VWIntLeaner` and `vw.learner.VWFloatLeaner`
2. Disallows `gz` and `bz2` protocols in VFS 1 and 2 URLs.
3. Disallows `vw.learner.VWIntArrayLeaner` and `vw.learner.VWFloatArrayLeaner`

Some of the scaladoc may not be correct in `VwJniModel`.  That probably needs to be updated.  Also, we should review the unit tests to make sure we can't do somewhat better in terms of refactoring.

### CODE REVIEW

**Comment by deaktator** [comment]
**Definitely need to bump version** to `3.0.0`.

File: aloha-core/src/main/scala/com/eharmony/aloha/io/sources/modelSources.scala [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  }
+
+  def shouldDelete = true
+
+  private[this] def copyToTmpFile(v: Vfs) = Try { v.fromByteArray(Base64.decodeBase64(b64EncodedData)) }
+
+  private[this] def getTmpVfs(tmpName: String) = Vfs.fromVfsType(vfsType)(tmpName)
+
+  private[this] def getTmpFileName = for {
+    sha1    <- Try { MessageDigest.getInstance("SHA-1") }
```
Comment (jmorra) [inline aloha-core/src/main/scala/com/eharmony/aloha/io/sources/modelSources.scala]: Can't this be memoized?

Also, can you please add a comment about why this is necessary and we can't just call `Base64.encodeBase64(...)` directly?

File: aloha-core/src/main/scala/com/eharmony/aloha/reflect/RefInfoOps.scala [hunk 2]
```diff
@@ -167,7 +168,16 @@ object RefInfoOps extends RefInfoOps[RefInfo] with EitherHelpers {
   @SuppressWarnings(Array("deprecation"))
   def isSubType[Sub, Super](implicit sub: RefInfo[Sub], sup: RefInfo[Super]): Boolean = sub <:< sup
 
+  def isJavaInterface[A](implicit a: RefInfo[A]) = a.erasure.isInterface
```
Comment (jmorra) [inline aloha-core/src/main/scala/com/eharmony/aloha/reflect/RefInfoOps.scala:171]: Why is this necessary now?

File: aloha-vw-jni/src/main/scala/com/eharmony/aloha/models/vw/jni/Cli.scala [hunk 3]
```diff
@@ -89,8 +89,8 @@ object Cli {
                 if (2 <= knots.size) Right(()) else Left("spline-knots must contain at least 2 knots.")
             } text "max value for spline domain. (must additional provide spline-min, spline-delta, and spline-knots)." optional() maxOccurs(1)
             checkConfig { c =>
-                FsInstance.fromFsType(c.fsType)(c.spec)
-                FsInstance.fromFsType(c.fsType)(c.model)
+//                Vfs.fromVfsType(c.vfsType)(c.spec)
```
Comment (jmorra) [inline aloha-vw-jni/src/main/scala/com/eharmony/aloha/models/vw/jni/Cli.scala]: Can we remove commented code please.

File: aloha-vw-jni/src/main/scala/com/eharmony/aloha/models/vw/jni/VwJniModel.scala [hunk 4]
```diff
@@ -20,25 +21,24 @@ import com.eharmony.aloha.score.basic.ModelOutput
...
 /**
  * Model that delegates to a VW JNI model.
  * @param modelId: ModelIdentity
- * @param vwJniModelSource either the FsInstance pointing to the VW model or the Base64 encoded VW model.
+ * @param modelSource either the FsInstance pointing to the VW model or the Base64 encoded VW model.
  * @param featureNames names of features (parallel to featureFunctions)
  * @param featureFunctions functions to extract values from the input value
  * @param defaultNs indices of features that will be placed in the default VW namespace
  * @param namespaces mapping from namespace name to indices of features that will be placed in the namespace
- * @param finalizer a function that transforms the native VW output type (Float) to B
```
Comment (jmorra) [inline aloha-vw-jni/src/main/scala/com/eharmony/aloha/models/vw/jni/VwJniModel.scala:41]: I think you're missing a comment about learnerCreator.

File: aloha-core/src/main/scala/com/eharmony/aloha/reflect/RefInfoOps.scala (hunk 2 above)
Comment (deaktator) [inline aloha-core/src/main/scala/com/eharmony/aloha/reflect/RefInfoOps.scala:171]: This is for h2o compiler.

File: aloha-core/src/main/scala/com/eharmony/aloha/io/sources/modelSources.scala (hunk 1 above)
Comment (deaktator) [inline aloha-core/src/main/scala/com/eharmony/aloha/io/sources/modelSources.scala]: This could be memorized. I just wanted to eliminate state. Do you have a preference?

File: aloha-vw-jni/src/main/scala/com/eharmony/aloha/models/vw/jni/Cli.scala (hunk 3 above)
Comment (deaktator) [inline aloha-vw-jni/src/main/scala/com/eharmony/aloha/models/vw/jni/Cli.scala]: I'll remove the commented code.

File: aloha-vw-jni/src/main/scala/com/eharmony/aloha/models/vw/jni/VwJniModel.scala (hunk 4 above)
Comment (deaktator) [inline aloha-vw-jni/src/main/scala/com/eharmony/aloha/models/vw/jni/VwJniModel.scala:41]: Yeah. I was thinking about this before. That's why I commented in the PR. I'll fix. Additionally, I'll add more of a comment in the scaladoc body about the reasoning.

File: aloha-vw-jni/src/main/scala/com/eharmony/aloha/models/vw/jni/VwJniModel.scala [hunk 5]
```diff
@@ -49,234 +49,721 @@ import scala.util.{Failure, Success, Try}
...
-                    semantics.createFunction[Iterable[(String, Double)]](spec, default).
-                        left.map { Seq(s"Error processing spec '$spec'") ++ _ }. // Add the spec that errored.
-                        right.map { f => (k, f) }
-            }
     }
 
-    def getConversionFunction[C : RefInfo, B : ScoreConverter](implicit riC: RefInfo[C], scB: ScoreConverter[B]) = {
-      TypeCoercion[C, B](riC, scB.ri) getOrElse {
-        throw new DeserializationException(s"Couldn't find conversion function to ${RefInfoOps.toString(scB.ri)}")
+    // TODO: Copied from RegressionModel.  Refactor for reuse.
```
Comment (jmorra) [inline aloha-vw-jni/src/main/scala/com/eharmony/aloha/models/vw/jni/VwJniModel.scala]: Is it worthwhile to address this todo now?

File: aloha-vw-jni/src/main/scala/com/eharmony/aloha/models/vw/jni/VwJniModel.scala [hunk 6]
```diff
@@ -49,234 +49,721 @@ import scala.util.{Failure, Success, Try}
...
+    // TODO: If this doesn't work, use commons-lang3 StringEscapeUtils.unescapeJava for unescaping.
+    //       Removed commons-lang3 as a dependency because it's only used in 2 places.  Here and
+    //       aloha-core CsvModelRunner class.  Here's how it was originally.
+    //
+    //         val vwParams = Option(vwArgs).filter(_.trim.nonEmpty).map(args => Right(StringEscapeUtils.escapeJson(args)))
+    val vwParams = vwArgs.filter(_.trim.nonEmpty).map(args => Right(escape(args)))
+
+    val params = vwArgs.getOrElse("").trim
+
+    //        val timeMs = System.currentTimeMillis()
```
Comment (jmorra) [inline aloha-vw-jni/src/main/scala/com/eharmony/aloha/models/vw/jni/VwJniModel.scala]: More commented code.

File: aloha-vw-jni/src/main/scala/com/eharmony/aloha/models/vw/jni/VwJniModel.scala [hunk 7]
```diff
@@ -49,234 +49,721 @@ import scala.util.{Failure, Success, Try}
...
+    else {
+      val b64Model = VwJniModel.readBinaryVwModelToB64String(model.inputStream)
+      Vw(Base64StringSource(b64Model, VfsType.vfs2), vwParams)
     }
+
+    VwJNIAst(VwJniModel.parser.modelType, id, features, vwObj, ns, numMissingThreshold, notes, spline, classLabels).toJson
+  }
+
+
+//  /**
```
Comment (jmorra) [inline aloha-vw-jni/src/main/scala/com/eharmony/aloha/models/vw/jni/VwJniModel.scala]: A whole bunch of commented code.

File: aloha-vw-jni/src/test/scala/com/eharmony/aloha/models/vw/jni/VwJniModelTest.scala [hunk 8]
```diff
@@ -79,29 +81,31 @@ class VwJniModelTest extends Logging {
...
-            fail("VW should throw a java.lang.Exception with message: \"option '--link' cannot be specified more than once\"");
-        }
-        catch {
-            // Can't rely on msg because different versions of boost (a dep of VW) return different messages
-            // case e: IllegalArgumentException if e.getClass.getSimpleName == "IllegalArgumentException" && e.getMessage == "option '--link' cannot be specified more than once" =>
-            case e: IllegalArgumentException if e.getClass.getSimpleName == "IllegalArgumentException" =>
-            case t: Throwable => throw t
-        }
+    @Ignore @Test def testFailureWhenVwIsCreatedWithLinkParamAndInitialRegressorHasSameLink(): Unit = {
+//        try {
```
Comment (jmorra) [inline aloha-vw-jni/src/test/scala/com/eharmony/aloha/models/vw/jni/VwJniModelTest.scala]: Commented code

Comment (deaktator) [inline aloha-vw-jni/src/test/scala/com/eharmony/aloha/models/vw/jni/VwJniModelTest.scala]: The reason I did this is that was for expediency.  Should I just remove the test?  I wanted to essentially make a placeholder saying gzipped and bzip2ed protocols don't work anymore.

File: aloha-vw-jni/src/test/scala/com/eharmony/aloha/models/vw/jni/VwJniModelTest.scala [hunk 9]
```diff
@@ -79,29 +81,31 @@ class VwJniModelTest extends Logging {
...
-    @Test def testAllocatedModelEqualsOriginalModel(): Unit = {
-        val modelBytes = readFile(VwModelFile)
-        val out = new String(Base64.encodeBase64(modelBytes))
-        val src = Base64EncodedBinaryVwModelSource(out)
-        val tmpFile = src.localFile(1)
-        src.copyContentToLocalIfNecessary(tmpFile)
-        val tmpBytes = readFile(tmpFile)
-        println(out)
-        assertArrayEquals(modelBytes, tmpBytes)
+    @Ignore @Test def testAllocatedModelEqualsOriginalModel(): Unit = {
```
Comment (jmorra) [inline aloha-vw-jni/src/test/scala/com/eharmony/aloha/models/vw/jni/VwJniModelTest.scala]: Commented code

File: aloha-vw-jni/src/test/scala/com/eharmony/aloha/models/vw/jni/VwJniModelTest.scala [hunk 10]
```diff
@@ -260,13 +264,19 @@ class VwJniModelTest extends Logging {
...
-        val params = tmp.vwJniModelSource.updatedVwModelParams(tmp.modelId.getId())
+        val file = """^(.*\s+)?-i\s+([^\s]*)(\s.*)?$""".r
+        val params = getParams(tmp)
+
+        assertFalse(tmp.modelSource.shouldDelete)
+
         params match {
-            case file(tmpFile) =>
-                assertFalse(s"Temp file ($tmpFile) should already be deleted.", new File(tmpFile).exists())
+            // This won't necessarily be deleted because we're not calling the same localModelFile invocation called
```
Comment (jmorra) [inline aloha-vw-jni/src/test/scala/com/eharmony/aloha/models/vw/jni/VwJniModelTest.scala:255]: Is this a problem because it'll leave garbage around after the tests are run?

File: aloha-vw-jni/src/test/scala/com/eharmony/aloha/models/vw/jni/VwJniModelTest.scala [hunk 11]
```diff
@@ -277,19 +287,24 @@ class VwJniModelTest extends Logging {
...
 
-        val file = """^.*\s+-i\s*([^\s]*\.model).*$""".r
-        val params = ram.vwJniModelSource.updatedVwModelParams(ram.modelId.getId())
+        val file = """^(.*\s+)?-i\s+([^\s]*)(\s.*)?$""".r
+        val params = getParams(ram)
 
         params match {
-            case file(tmpFile) =>
-                assertFalse(s"Temp file ($tmpFile) should already be deleted.", new File(tmpFile).exists())
+            // This won't necessarily be deleted because we're not calling the same localModelFile invocation called
```
Comment (jmorra) [inline aloha-vw-jni/src/test/scala/com/eharmony/aloha/models/vw/jni/VwJniModelTest.scala:275]: Same as above

File: aloha-vw-jni/src/test/scala/com/eharmony/aloha/models/vw/jni/VwJniModelTest.scala [hunk 12]
```diff
@@ -300,20 +315,29 @@ class VwJniModelTest extends Logging {
...
+     */
+    def assertBadFormat[A](a: => A): Unit = {
+        try {
+            a
+            fail()
+        }
+        catch {
+            case e: Throwable if "bad model format!" == e.getMessage =>
+                val w = new StringWriter()
+                e.printStackTrace(new PrintWriter(w))
```
Comment (jmorra) [inline aloha-vw-jni/src/test/scala/com/eharmony/aloha/models/vw/jni/VwJniModelTest.scala:316]: `printStackTrace`?

File: aloha-vw-jni/src/test/scala/com/eharmony/aloha/models/vw/jni/VwJniModelTest.scala [hunk 13]
```diff
@@ -323,18 +347,20 @@ class VwJniModelTest extends Logging {
...
-
-        params match {
-            case file(tmpFile) =>
-                assertFalse(s"Temp file ($tmpFile) should already be deleted.", new File(tmpFile).exists())
-                assertTrue("'zip' URLs should copy VW model to temp directory.", params.contains(TmpDir))
-            case _ => fail("Should have a temp file location")
-        }
+        assertBadFormat(model[Float](extJson(bz2Url.toString)))
+
+//        // Show that we can take zipped URLs and that they are copied to a local temp file.
```
Comment (jmorra) [inline aloha-vw-jni/src/test/scala/com/eharmony/aloha/models/vw/jni/VwJniModelTest.scala]: Commented code

File: aloha-core/src/main/scala/com/eharmony/aloha/io/sources/modelSources.scala (hunk 1 above)
Comment (jmorra) [inline aloha-core/src/main/scala/com/eharmony/aloha/io/sources/modelSources.scala]: If it's in a companion object it'll still be stateless.

File: aloha-core/src/main/scala/com/eharmony/aloha/reflect/RefInfoOps.scala (hunk 2 above)
Comment (jmorra) [inline aloha-core/src/main/scala/com/eharmony/aloha/reflect/RefInfoOps.scala:171]: OK

File: aloha-vw-jni/src/main/scala/com/eharmony/aloha/models/vw/jni/VwJniModel.scala (hunk 4 above)
Comment (jmorra) [inline aloha-vw-jni/src/main/scala/com/eharmony/aloha/models/vw/jni/VwJniModel.scala:41]: Sounds good.  Do you not like the interface?  I think it's pretty elegant given the restriction that `VWLearner` cannot be serialized.  That's worth a comment.

File: aloha-vw-jni/src/test/scala/com/eharmony/aloha/models/vw/jni/VwJniModelTest.scala (hunk 8 above)
Comment (jmorra) [inline aloha-vw-jni/src/test/scala/com/eharmony/aloha/models/vw/jni/VwJniModelTest.scala]: Leave the code uncommented with an `@Ignore` so that we know it still compiles, and then just add a comment about why it's ignored.

Comment (deaktator) [inline aloha-vw-jni/src/test/scala/com/eharmony/aloha/models/vw/jni/VwJniModelTest.scala]: Whoops.  Those were commented out because they don't really belong here.  Those should go in vw jni library.

**Comment by jmorra** [comment]
Once you address my comments I think this looks really good.  I think the best way to proceed is to merge to master and then I'll run some tests on my side to verify everything still works.  Once I see it's all good let's go ahead and release.

**Comment by deaktator** [comment]
Not sure what's going on.  I am getting permgen OOM exceptions during compilation.  Trying to fix now

**Comment by jmorra** [comment]
@deaktator I just compared one of our models that uses VW using the currently released Aloha and this branch over 8962 protos and both gave the same results, so I feel good about this.

**Comment by deaktator** [comment]
@jmorra Awesome.  Question here.  aloha-core was failing to compile with a PermGen out of memory error.  It compiles fine with zinc for some reason.  Should I just try to commit without changes to the POM and see if travis fails, and if so, bump the permgen?

**Comment by jmorra** [comment]
I think that is a good idea. I was compiling with Java 8 so that wasn't an
issue for me.

**Comment by deaktator** [comment]
OK.  Travis passed.  Take a look at the new stuff and let me know what you think.

File: aloha-vw-jni/src/test/scala/com/eharmony/aloha/models/vw/jni/VwJniModelTest.scala [hunk 14]
```diff
@@ -131,6 +107,16 @@ class VwJniModelTest extends Logging {
         assertEquals(List("height_cm"), m.featureFunctions.head.accessorOutputMissing(missingHeight))
     }
 
+  @Ignore @Test def testAllocatedModelEqualsOriginalModel(): Unit = {
```
Comment (jmorra) [inline aloha-vw-jni/src/test/scala/com/eharmony/aloha/models/vw/jni/VwJniModelTest.scala]: Comment about why ignored.

File: aloha-core/src/main/scala/com/eharmony/aloha/io/sources/modelSources.scala [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+
+  /**
+   * This is based on a URL-encoded URL of the base64 encoding of a hash of the data.  The reason this is
+   * necessary is that we want a short, safe URL description of the content.
+   *
+   * NOTE: It's possible that hash collisions could occur.  This must be dealt with by the calling code.
+   * @return
+   */
+  private[this] def getTmpFileName = for {
+    sha1    <- Try { MessageDigest.getInstance("SHA-1") }
```
Comment (jmorra) [inline aloha-core/src/main/scala/com/eharmony/aloha/io/sources/modelSources.scala]: Memoize?

**Comment by jmorra** [comment]
Added 2 comments and then still have a few that are unaddressed in the tests, specifically about cleaning up from the tests.

**Comment by deaktator** [comment]
@jmorra Not sure exactly what else you want.  I think I committed the two changes just now.  Let me know what you think.

**Comment by jmorra** [comment]
I think it looks fine now, let's go ahead and merge and bump up the version in the pom to 3.0.0-SNAPSHOT.
